js/battle.js

```diff
@@ -5679,12 +5680,14 @@ var Battle = (function () {
 			break;
 		case 'join':
 		case 'j':
+			this.users[toUserid(args[1])] = ' ' + args[1];
 			if (!this.ignoreSpects) {
 				this.log('<div class="chat"><small>' + Tools.escapeHTML(args[1]) + ' joined.</small></div>', preempt);
 			}
 			break;
 		case 'leave':
 		case 'l':
+			delete this.users[toUserid(args[1])];
```

Comment:
To answer your questions in order:
1. No, you don't need to delete old users, but the chat rooms do (a room mod asked me about this because he wanted to tab complete hit and run spammers)
2. Battles don't seem to have rename events.
3. I made the room take a reference to the battle because not all battles have rooms, i.e. replays, so it was easier that way around.